### PR TITLE
Add metropolis exchange rule for fermions

### PR DESCRIPTION
**Pull request by jwnys**
This sampler avoids proposing the same sample for fermions.
The previous approach was to use ExchangeSampler, which often applies an exchange between two occupied or two unoccupied fermionic modes.
Now we only consider clusters of one occupied and one unoccupied mode.
This should yield better acceptances.
Relevant for @inailuig and @imi-hub , and most specifically for the few-fermion case

### CODE REVIEW

File: Examples/Fermions/fermi_hubbard.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def nc(site, sz):
 
 # g.n_nodes == L*L --> disj_graph == 2*L*L
 disj_graph = nk.graph.disjoint_union(g, g)
-sa = nk.sampler.MetropolisExchange(hi, graph=disj_graph, n_chains=16)
+sa = nkx.sampler.MetropolisFermionExchange(hi, graph=disj_graph, n_chains=16)
```
Comment (gcarleo) [inline Examples/Fermions/fermi_hubbard.py:57]: since we are doing a specific sampler for fermions (which is very much needed, thanks!) why don't we make it such we don't need to do anymore this hack of the disjoint union?

File: netket/experimental/sampler/rules/fermion_2nd.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+
+class FermionExchangeRule(ExchangeRule):
+    """Exchange rule for fermions.
+
+    Works similarly to exchange rule, but:
+    takes into account that only occupied orbitals
+    can be exchanged with unoccupied ones.
+    """
+
```
Comment (PhilipVinc) [inline netket/experimental/sampler/rules/fermion_2nd.py:13]: I would include this in here, as it will be the standard use-case.

We can also ask the Hilbert space in here in order to make sure we get the right number of spin values... I dont know if this makes sense, but it seems to me it might?

```suggestion

    def __init__(
        self,
        hilbert, 
        *,
        clusters: Optional[list[tuple[int, int]]] = None,
        graph: Optional[AbstractGraph] = None,
        d_max: int = 1,
        copy_per_spin: bool = True,
    ):
        r"""
        Constructs the Exchange Rule.

        You can pass either a list of clusters or a netket graph object to
        determine the clusters to exchange.

        Args:
            clusters: The list of clusters that can be exchanged. This should be
                a list of 2-tuples containing two integers. Every tuple is an edge,
                or cluster of sites to be exchanged.
            graph: A graph, from which the edges determine the clusters
                that can be exchanged.
            d_max: Only valid if a graph is passed in. The maximum distance
                between two sites
            copy_per_spin: (default True) If True, the graph must encode the
                connectivity  between the first N physical sites having same spin, and
                it is replicated using :ref:`netket.graph.disjoint_union` other every
                spin subsector. This option conserves the number of fermions per
                spin subsector. 
        """
        if copy_per_spin and hilbert.n_spin_subsectors > 1:
            if graph is not None and graph.n_nodes == hilbert.n_orbitals:
                 graph = disjoint_union(*[graph] * hilbert.n_spin_subsectors)
            if clusters is not None and np.max(clusters) < hilbert.n_orbitals:
                clusters = np.concatenate(
                    [
                        clusters + i * hilbert.n_orbitals
                        for i in range(hilbert.n_spin_subsectors)
                    ]
                )
        super().__init__(clusters=clusters, graph=graph, d_max=d_max) 

```

**Review by PhilipVinc — COMMENTED** [review]
My only doubt is that maybe it would be nice to keep the spin flag in the rule itself. 
The samplers object are usually just wrappers for writing more Terse code...

what do you think?

File: netket/experimental/sampler/metropolis.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+from netket.sampler import MetropolisSampler
+from netket.experimental.hilbert import SpinOrbitalFermions
+from netket.graph import disjoint_union
+import numpy as np
+
+
+def MetropolisFermionExchange(
+    hilbert, *, clusters=None, graph=None, d_max=1, copy_per_spin=True, **kwargs
```
[the remark this answers is not in the document]
Comment (jwnys) [reply netket/experimental/sampler/metropolis.py:8]: I'm open to suggestions for the naming.

**Comment by jwnys** [comment]
Two main changes:
- change default dtype to integers (exchange gives float64).
- I renamed the sampler to a more general "ParticleExchange". After talking to @Z-Denis , the same sampler can be used for bosons as well (after generalizing the way hoppable_clusters are identified). I didn't allow for bosons yet, because it seems that there is no such thing as SpinOrbitalBosons, or similar. I'll leave this up to @Z-Denis  for bosons. Nothing needs to be changed, it's just not clear to me what to do with the spin checks for that. What do you think @PhilipVinc @gcarleo ?

**Review by PhilipVinc — APPROVED** [review]
> change default dtype to integers (exchange gives float64).

I'm all in favour for this change. I actually think we should use this everywhere...

I'd just like to point out that netket's operators (and everything else) is not throughly tested for integer sample dtypes, so be on the lookout for problems that might appear elsewhere, like QGT.

> I renamed the sampler to a more general "ParticleExchange". 

Good idea! I have the code to do this for bosons somewhere... It applies to Spin->=1 spaces as well.

> it's just not clear to me what to do with the spin checks for that

can be done in a follow up PR by someone else.

--

So the only remaining thing for me is whether we keep `copy_per_spin` or rename it. Otherwise this can be merged.

File: netket/experimental/sampler/rules/fermion_2nd.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    def __init__(
+        self,
+        hilbert,
+        *,
+        clusters: Optional[list[tuple[int, int]]] = None,
+        graph: Optional[AbstractGraph] = None,
+        d_max: int = 1,
+        copy_per_spin: bool = True,
```
Comment (PhilipVinc) [inline netket/experimental/sampler/rules/fermion_2nd.py:35]: Actually @jwnys , would it make sense to default this to `None` and have it default to
```python
if copy_per_spin is None:
    copy_per_spin = all(n is not None for n in hilbert.n_fermions_per_spin)
```
?

**Comment by gcarleo** [comment]
Yes I also think copy is not the right word here, can we have something
like swap_spins or similar?

**Comment by PhilipVinc** [comment]
Same spin, rather...

**Comment by jwnys** [comment]
Hm, then I'd rather like to keep it more explicit that we are copying the interaction graph per spin sector..

File: netket/experimental/sampler/rules/fermion_2nd.py [hunk 5]
```diff
@@ -1,15 +1,79 @@
...
     can be exchanged with unoccupied ones.
+    This sampler conserves the number of particles.
+    """
+
+    copy_per_spin: bool
     """
+    Whether to copy the graph for each spin sector.
+    This only occurs if cluster indices or number of graph nodes
+    are lower than the number of orbitals per spin sector.
+    """
```
[the remark this answers is not in the document]
Comment (jwnys) [reply netket/experimental/sampler/rules/fermion_2nd.py]: You're right, I'll remove it

**Comment by jwnys** [comment]
Maybe "spin_symmetric" then?

**Comment by PhilipVinc** [comment]
But is this simmetry? isn't it more invariant? Maybe `replicate_spin_subsectors` or `identical_spin_subsectors` ?

**Comment by jwnys** [comment]
let's maybe just keep the copy for now then?

**Comment by gcarleo** [comment]
What does the flag do exactly? It disallows exchanging different spins so that the spin populations are constant ? If so, I'd not use copy it's really not clear what it means. Just call it exhange_spins=False or something?

**Comment by jwnys** [comment]
> exhange_spins

Yeah, it's just that I rather copy the graph per spin, so I'm not removing any spin exchanges if you were to pass a graph that would do that.

**Comment by gcarleo** [comment]
I see... Then maybe can we impose the graph size is the same as the number of sites? (not twice as much, which is a hack and should not be used)

**Comment by jwnys** [comment]
> I see... Then maybe can we impose the graph size is the same as the number of sites? (not twice as much, which is a hack and should not be used)

Alternatively, I just mention explicitly in the docs that i this flag only has an impact if the number of sites = number of orbitals..

**Comment by PhilipVinc** [comment]
> Yeah, it's just that I rather copy the graph per spin, so I'm not removing any spin exchanges if you were to pass a graph that would do that

Though right now it would crash, because you'd end up with a graph with more connections.
Crashing is kind-of like saying it does not work in other cases.

> exhange_spins

I like this. It could be that it can be used only if you have conserved number per spin, and you cannot specify it otherwise...

File: netket/experimental/sampler/metropolis.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    **kwargs,
+) -> MetropolisSampler:
+    r"""
+    This sampler moves (or hops) a random particle to a different but random empty mode.
+    It works similar to MetropolisExchange, but only allows exchanges between occupied and unoccupied modes.
+
+    Args:
+        hilbert: The Hilbert space to sample.
+        d_max: The maximum graph distance allowed for exchanges.
+        exhange_spins: If False, exchanges are only allowed between modes with the same spin projection.
```
Comment (PhilipVinc) [inline netket/experimental/sampler/metropolis.py:22]:
```suggestion
        exhange_spins: (default False) If False, exchanges are only allowed between modes with the same spin projection.
```

File: netket/experimental/sampler/rules/fermion_2nd.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+from netket.sampler.rules import ExchangeRule
+from netket.graph import AbstractGraph
+from netket.graph import disjoint_union
+from netket.experimental.hilbert import SpinOrbitalFermions
+
+
+class ParticleExchangeRule(ExchangeRule):
+    """Exchange rule for particles on a lattice.
+
+    Works similarly to ExchangeRule, but:
```
Comment (PhilipVinc) [inline netket/experimental/sampler/rules/fermion_2nd.py:15]:
```suggestion
    Works similarly to :class:`netket.sampler.rules.ExchangeRule`, but:
```

File: netket/experimental/sampler/rules/fermion_2nd.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    takes into account that only occupied orbitals
+    can be exchanged with unoccupied ones.
+    This sampler conserves the number of particles.
```
Comment (PhilipVinc) [inline netket/experimental/sampler/rules/fermion_2nd.py:18]:
```suggestion

    This sampler conserves the number of particles.
```

File: docs/api/sampler.md [hunk 9]
```diff
@@ -83,6 +83,7 @@ This is a list of shorthands that allow to construct a {class}`~netket.sampler.M
 
    MetropolisLocal
    MetropolisExchange
+   MetropolisParticleExchange
```
Comment (PhilipVinc) [inline docs/api/sampler.md:86]: those go in `api/experimental.md`, not here. this will break the docs build

File: docs/api/sampler.md [hunk 10]
```diff
@@ -117,6 +118,7 @@ Sampler. Rules with `Numpy` in their name can only be used with
   rules.LocalRule
   rules.CustomRuleNumpy
   rules.ExchangeRule
+  rules.ParticleExchangeRule
```
Comment (PhilipVinc) [inline docs/api/sampler.md:121]: You could also add below a block like the following, which duplicates a bit the docs but maybe would help searchability

```
## Experimental models

The following samplers are experimental, meaning that we could change them at some point, and we actively seeking for feedback and opinions on their usage and APIs.


### Particle-specific samplers

The following models are for 2nd-quantisation fermionic hilbert spaces ({class}`netket.experimental.hilbert.SpinOrbitalFermions`).

```{eval-rst}
.. currentmodule:: netket.experimental.sampler

```

```{eval-rst}
.. autosummary::
   :toctree: _generated/samplers
   :template: flax_module_or_default
   :nosignatures:


   MetropolisParticleExchange
```

And the corresponding rules
```{eval-rst}
.. autosummary::
   :toctree: _generated/samplers
   :template: flax_module_or_default
   :nosignatures:


   rules.ParticleExchangeRule
```

File: netket/experimental/sampler/rules/fermion_2nd.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        return f"ParticleExchangeRule(# of clusters: {len(self.clusters)})"
+
+
+@jax.jit
+def _compute_hoppable_clusters_mask(clusters, σ):
+    # mask the clusters to include only feasible moves (occ -> unocc, or the inverse)
+    hoppable_clusters_mask = ~jnp.isclose(
+        σ[..., clusters[:, 0]], σ[..., clusters[:, 1]]
+    )
+    return hoppable_clusters_mask
```
Comment (PhilipVinc) [inline netket/experimental/sampler/rules/fermion_2nd.py:119]: Now that I think about it, this condition makes sense for Exchange Rule itself, because it simply asks that it does not swap two identical configurations.

Why can't this be used in the standard Exchange Rule? (cc @gcarleo )

Comment (PhilipVinc) [reply netket/experimental/sampler/rules/fermion_2nd.py:119]: Ok.. Maybe we can do that in a follow up PR.
